src/server/debug/server/server.go

```diff
@@ -325,16 +326,18 @@ func writeProfile(ctx context.Context, w io.Writer, profile *debug.Profile) erro
 				return errors.EnsureStack(err)
 			}
 		}
-
-		// Wait for either the defined duration, or until the context is done.  We'll eat
-		// the context cancelled / deadline exceeded error here, but I think that's fine
-		// since we don't do any other IO in this function.  (The writer is going to be
-		// written to until StopCPUProfile anyway.)
-		tctx, c := context.WithTimeout(ctx, duration)
-		<-tctx.Done()
-		c()
-		pprof.StopCPUProfile()
-		return nil
+		// Wait for either the defined duration, or until the context is
+		// done.
+		t := time.NewTimer(duration)
+		select {
+		case <-ctx.Done():
+			if !t.Stop() {
+				<-t.C
+			}
+			return errors.EnsureStack(ctx.Err())
+		case <-t.C:
+			return nil
+		}
```

Comment:
Why do it this way?  There are now 3 branches to test; ctx.Done + !t.Stop, ctx.Done + t.Stop, and recv on t.C.

Reply:
In order to distinguish between a cancelled context and an exceeded deadline (returning an error in the former situation but no error in the latter).

I think maybe t.Stop() can just be called and its return value be ignored, to be honest, because we’re not going to reuse the channel.